DatadogInternal/Sources/NetworkInstrumentation/SpanID.swift

```diff
@@ -0,0 +1,161 @@
+    /// - Parameter value: The value to create.
+    public init(integerLiteral value: UInt64) {
+        self.init(rawValue: value)
+    }
+}
+
+extension String {
+    /// Creates a `String` representation of a `SpanID`.
+    ///
+    /// - Parameters:
+    ///   - spanID: The Trace ID
+    ///   - representation: The required representation. `.decimal` by default.
+    public init(_ spanID: SpanID, representation: SpanID.Representation = .decimal) {
+        switch representation {
+        case .decimal:
+            self.init(spanID.rawValue)
+        case .hexadecimal:
+            self.init(spanID.rawValue, radix: 16)
+        case .hexadecimal16Chars:
+            self.init(format: "%016llx", spanID.rawValue)
+        case .hexadecimal32Chars:
+            self.init(format: "%032llx", spanID.rawValue)
+        }
+    }
+}
+
+/// A `SpanID` generator interface.
+public protocol SpanIDGenerator {
+    /// Generates a new and unique `SpanID`.
+    ///
+    /// - Returns: The generated `SpanID`
+    func generate() -> SpanID
+}
+
+/// A Default `SpanID` genarator.
+public struct DefaultSpanIDGenerator: SpanIDGenerator {
+    /// Describes the lower and upper boundary of tracing ID generation.
+    ///
+    /// * Lower: starts with `1` as `0` is reserved for historical reason: 0 == "unset", ref: dd-trace-java:DDId.java.
+    /// * Upper: equals to `2 ^ 63 - 1` as some tracers can't handle the `2 ^ 64 -1` range, ref: dd-trace-java:DDId.java.
+    public static let defaultGenerationRange = (1...UInt64.max >> 1)
+
+    /// The generator's range.
+    let range: ClosedRange<UInt64>
+
+    /// Creates a default generator.
+    ///
+    /// - Parameter range: The generator's range.
+    public init(range: ClosedRange<UInt64> = Self.defaultGenerationRange) {
+        self.range = range
+    }
+
+    /// Generates a new and unique `SpanID`.
+    ///
+    /// The Trace ID will be generated within the range.
+    ///
+    /// - Returns: The generated `SpanID`
+    public func generate() -> SpanID {
+        var rawValue: UInt64
+        repeat {
+            rawValue = UInt64.random(in: range)
+        } while rawValue == SpanID.invalidId
+        return SpanID(rawValue: rawValue)
+   }
+}
+
+extension SpanID: Codable {
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        do {
+            let rawValue = try container.decode(UInt64.self)
+            self.init(rawValue: rawValue)
+        } catch {
+            let rawValue = try container.decode(String.self)
+            guard let spanID = SpanID(rawValue, representation: .decimal) else {
+                throw DecodingError.dataCorruptedError(in: container, debugDescription: "Invalid SpanID format: \(rawValue)")
+            }
+            self = spanID
+        }
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        try container.encode(rawValue)
+    }
+}
```

Comment:
suggestion/ We're missing unit tests coverage for `SpanID` coding (in `DatadogInternalTests`). The overall coverage for `SpanID` looks quite low given the complexity we manage here - could we improve it?

<img width="743" alt="Screenshot 2024-04-09 at 10 29 02" src="https://github.com/DataDog/dd-sdk-ios/assets/2358722/fe823cd3-e3af-45e6-98b6-1e9a9c22cef5">

Reply:
fair call, I added bunch of cases now.